dff/utils/benchmark/context_storage.py

```diff
@@ -0,0 +1,311 @@
+    as_dataframe: None = None,
+) -> tp.Tuple[tp.List[float], tp.List[float]]:
+    ...
+
+
+@tp.overload
+def time_context_read_write(
+    context_storage: DBContextStorage,
+    context: Context,
+    context_num: int,
+    as_dataframe: tp.Literal["pandas"],
+) -> "pandas.DataFrame":
+    ...
+
+
+@tp.overload
+def time_context_read_write(
+    context_storage: DBContextStorage,
+    context: Context,
+    context_num: int,
+    as_dataframe: tp.Literal["polars"],
+) -> "polars.DataFrame":
+    ...
+
+
+def time_context_read_write(
+    context_storage: DBContextStorage,
+    context: Context,
+    context_num: int,
+    as_dataframe: tp.Optional[tp.Literal["pandas", "polars"]] = None,
+) -> tp.Union[tp.Tuple[tp.List[float], tp.List[float]], "pandas.DataFrame", "polars.DataFrame"]:
+    """
+    Generate `context_num` ids and for each write into `context_storage` value of `context` under generated id,
+    after that read the value stored in `context_storage` under generated id and compare it to `context`.
+
+    Keep track of the time it takes to write and read context to/from the context storage.
+
+    This function clear context storage before and after execution.
+
+    :param context_storage: Context storage to benchmark.
+    :param context: An instance of context which will be repeatedly written into context storage.
+    :param context_num: A number of times the context will be written and checked.
+    :param as_dataframe:
+        If the function should return the results as a pandas or a polars DataFrame.
+        If set to None, does not return a Dataframe.
+        Defaults to None.
+    :return:
+        Depends on `as_dataframe` parameter.
+        1. By default, it is set to None in which case it returns:
+            two lists: first one contains individual write times, second one contains individual read times.
+        2. If set to "pandas":
+            A pandas DataFrame with two columns: "write" and "read" which contain corresponding data series.
+        3. If set to "polars":
+            A polars DataFrame with the same columns as in a pandas DataFrame.
+    :raises RuntimeError: If context written into context storage does not match read context.
+    """
+    context_storage.clear()
+
+    write_times: tp.List[float] = []
+    read_times: tp.List[float] = []
+    for _ in tqdm(range(context_num), desc=f"Benchmarking context storage:{context_storage.full_path}"):
+        ctx_id = uuid4()
+
+        # write operation benchmark
+        write_start = perf_counter()
+        context_storage[ctx_id] = context
+        write_times.append(perf_counter() - write_start)
+
+        # read operation benchmark
+        read_start = perf_counter()
+        actual_context = context_storage[ctx_id]
+        read_times.append(perf_counter() - read_start)
+
+        # check returned context
+        if actual_context != context:
+            raise RuntimeError(f"True context:\n{context}\nActual context:\n{actual_context}")
```

Comment:
This will never succeed if we read, say, 3 last requests from context storage only.
We should mind that OR manually set read/write policy of context storage to ALL.
But that should be done after merge of partial context updates only.